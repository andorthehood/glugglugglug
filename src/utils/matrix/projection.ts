/**
 * Builds an orthographic projection matrix that maps the unit cube to clip space.
 * @param width Scene width.
 * @param height Scene height (positive values map upward).
 * @param depth Depth of the viewing volume.
 * @returns A 4x4 orthographic projection matrix.
 */
export function projection(width: number, height: number, depth: number) {
	return [
		2 / width,
		0,
		0,
		0,
		0,
		-2 / height,
		0,
		0,
		0,
		0,
		2 / depth,
		0,
		-1,
		1,
		0,
		1,
	];
}
