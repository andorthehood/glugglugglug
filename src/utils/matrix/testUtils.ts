export function roundValue(value: number, precision = 6) {
	const factor = 10 ** precision;
	const rounded = Math.round(value * factor) / factor;
	return Object.is(rounded, -0) ? 0 : rounded;
}

export function formatMatrix(values: number[]) {
	return Array.from({ length: 4 }, (_, row) => {
		const start = row * 4;
		return values.slice(start, start + 4).map((value) => roundValue(value));
	});
}

export function formatVector(values: number[]) {
	return values.map((value) => roundValue(value));
}
