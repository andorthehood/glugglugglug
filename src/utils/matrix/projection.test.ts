import { describe, expect, it } from "vitest";
import { projection } from "./projection";
import { formatMatrix } from "./testUtils";

describe("projection", () => {
	it("builds an orthographic projection matrix with the current implementation", () => {
		const result = projection(4, 2, 8);

		expect(formatMatrix(result)).toMatchSnapshot();
	});
});
