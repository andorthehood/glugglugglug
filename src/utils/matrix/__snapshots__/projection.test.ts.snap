// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`projection > builds an orthographic projection matrix with the current implementation 1`] = `
[
  [
    0.5,
    0,
    0,
    0,
  ],
  [
    0,
    -1,
    0,
    0,
  ],
  [
    0,
    0,
    0.25,
    0,
  ],
  [
    -1,
    1,
    0,
    1,
  ],
]
`;
